File: 3170.lexicographically-minimum-string-after-removing-stars.py
#
# @lc app=leetcode id=3170 lang=python3
#
# [3170] Lexicographically Minimum String After Removing Stars
#

# @lc code=start
from itertools import chain
from sortedcontainers import SortedDict
class Solution:
    def clearStars(self, s: str) -> str:
        #track the smallest non star character on the left of the 
        mp = SortedDict()  

        #Record the index of elements to be deleted
        st = set()
        
        for i, c in enumerate(s):
            if c == '*':
                key, indices = mp.peekitem(0)
                st.add(indices.pop())
                if not indices:
                    del mp[key]
            else:
                if c not in mp:
                    mp[c] = []
                mp[c].append(i)
        ans_chars = [c for i, c in enumerate(s) if i not in st and c != '*']
        return ''.join(ans_chars)

        #todo use 26 stacks to store the index of each of the characters
        # stacks = [[] for _ in range(26)]
        # for i, c in enumerate(s):
        #     if c != '*':
        #         stacks[ord(c) - ord('a')].append(i)
        #         continue
        #     for stack in stacks:
        #         if stack:
        #             stack.pop()
        #             break
        # return ''.join(s[i] for i in sorted(chain.from_iterable(stacks)))
# @lc code=end

File: 3170.lexicographically-minimum-string-after-removing-stars.java
/*
 * @lc app=leetcode id=3170 lang=java
 *
 * [3170] Lexicographically Minimum String After Removing Stars
 */

// @lc code=start
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Arrays;
class Solution {
    public String clearStars(String s) {
        //clarification
        //string, goal is to remove the *
        //remove two things: leftmost * and smallest non * character on the left
        //If we delete from left to right, delete the * when we meet one and the smallest non * character
        
        //! we are trying to keep the result lexicographically smallest
        //we also want to delete the closest one
        //"a-----a----*" by induction proof method, the condition holds
        Deque<Integer>[] stacks = new ArrayDeque[26];
        Arrays.setAll(stacks, i -> new ArrayDeque<>());
        char[] cArray = s.toCharArray();
        for(int i = 0; i < cArray.length; i++){
            if(cArray[i] == '*'){
                for(int j = 0; j < 26; j++){
                    if(!stacks[j].isEmpty()){
                        stacks[j].pop();
                        break;
                    }
                }
            }
            else{
                stacks[cArray[i] - 'a'].push(i);
            }
        }
        char[] c = new char[cArray.length];
        for(int i = 0; i < stacks.length; i++){
            while(!stacks[i].isEmpty()){
                c[stacks[i].pop()] = (char) ('a' + i);
            }
        }
        StringBuilder ans = new StringBuilder();
        for(int i = 0; i < c.length; i++){
            if(c[i] != 0){
                ans.append(c[i]);
            }
        }
        return ans.toString();
    }
}
// @lc code=end

